src/transformers/utils/fx.py

```diff
@@ -121,107 +122,313 @@ def _generate_supported_model_classes(
+
+
+def torch_index_select(input, dim, index, *, out=None):
+    shape = list(input.shape)
+    shape[dim] = len(index)
+    return torch.empty(*shape, device="meta")
+
+
+def torch_tensor_index_select(self, dim, index):
+    return torch_tensor_index_select(self, dim, index)
+
+
+def torch_nn_mseloss(self, input, target):
+    if self.reduction == "none":
+        shape = target.shape
+    else:
+        shape = (1,)
+    return torch.empty(shape, device="meta")
+
+
+def torch_nn_crossentropyloss(self, input, target):
+    if self.reduction == "none":
+        shape = target.shape
+    else:
+        shape = (1,)
+    return torch.empty(shape, device="meta")
+
+
+def torch_nn_bcewithlogitsloss(self, input, target):
+    if self.reduction == "none":
+        shape = target.shape
+    else:
+        shape = (1,)
+    return torch.empty(shape, device="meta")
+
+
+_MANUAL_META_OVERRIDES: Dict[Callable, Callable] = {
+    torch.nn.Embedding: embedding_override,
+    torch.nn.LayerNorm: torch_nn_layernorm_override,
+    torch.nn.Linear: torch_nn_linear_override,
+    torch.relu: torch_relu_override,
+    torch.nn.functional.relu: torch_nn_functional_relu_override,
+    torch.nn.ReLU: torch_nn_relu_override,
+    torch.where: torch_where_override,
+    torch.abs: torch_abs_override,
+    torch.arange: torch_arange_override,
+    torch.cat: torch_cat_override,
+    torch.stack: torch_stack_override,
+    torch.add: torch_add_override,
+    torch.mul: torch_mul_override,
+    torch.Tensor.mul: torch_tensor_mul_override,
+    torch.matmul: torch_matmul_override,
+    torch.Tensor.repeat: torch_tensor_repeat_override,
+    # TODO: those might not be needed.
+    # torch.index_select: torch_index_select,
+    # torch.Tensor.index_select: torch_tensor_index_select,
+    torch.nn.MSELoss: torch_nn_mseloss,
+    torch.nn.CrossEntropyLoss: torch_nn_crossentropyloss,
+    torch.nn.BCEWithLogitsLoss: torch_nn_bcewithlogitsloss,
+}
+
+
 class HFProxy(Proxy):
     """
-    Proxy that is able to provide the proper ranks, shapes and boolean values during symbolic tracing by implementing
-    the dim, size and __bool__ methods. It can be easily extended by either adding new methods or extending the
-    existing ones.
+    Proxy that uses metadata to handle data-dependent control-flow.
     """
 
-    def __init__(self, node: Node, tracer: Optional[Tracer] = None):
-        super().__init__(node, tracer=tracer)
-        if hasattr(self, "tracer") and self.tracer is not None:
-            self.device = self.tracer.root.device
-            self.dtype = next(self.tracer.root.parameters()).dtype
-            self.cache = None
+    def install_metadata(self, metadata):
+        self._metadata = metadata
 
     @property
     def shape(self):
-        return self.size()
-
-    def __setitem__(self, key, value):
-        pass
-
-    def __contains__(self, key):
-        return False
+        return self.tracer.create_proxy("call_method", "size", (self,), {})
```

Comment:
It seems like we cannot get concrete shape information with this implementation. I've tried, for example, `GPT2ModelTest::test_torch_fx` and it does not trace properly. Implementing shape and size as the following works:

```
    @property
    def shape(self):
        if hasattr(self, '_tensor_meta') and self._tensor_meta is not None:
            return self._tensor_meta.shape
        return self.tracer.create_proxy('call_function', builtins.getattr, (self, 'shape'), {})

    def size(self, dim=None):
        if hasattr(self, '_tensor_meta') and self._tensor_meta is not None:
            return self._tensor_meta.size(*[dim] if dim else [])
        return self.tracer.create_proxy('call_method', 'size', (self, dim) if dim else (self,), {})
```